modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/incoming/IncomingSnapshotCopier.java

```diff
@@ -0,0 +1,117 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table.distributed.raft.snapshot.incoming;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import org.apache.ignite.internal.table.distributed.raft.snapshot.PartitionSnapshotStorage;
+import org.apache.ignite.internal.table.distributed.raft.snapshot.SnapshotUri;
+import org.apache.ignite.internal.table.message.SnapshotMetaRequest;
+import org.apache.ignite.internal.table.message.SnapshotMetaResponse;
+import org.apache.ignite.internal.table.message.TableMessagesFactory;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.network.MessagingService;
+import org.apache.ignite.network.NetworkMessage;
+import org.apache.ignite.raft.jraft.entity.RaftOutter.SnapshotMeta;
+import org.apache.ignite.raft.jraft.storage.SnapshotStorage;
+import org.apache.ignite.raft.jraft.storage.snapshot.SnapshotCopier;
+import org.apache.ignite.raft.jraft.storage.snapshot.SnapshotReader;
+
+/**
+ * Snapshot copier implementation for partitions. Used to stream partition data from the leader to the local node.
+ */
+public class IncomingSnapshotCopier extends SnapshotCopier {
+    /** Messages factory. */
+    private static final TableMessagesFactory MSG_FACTORY = new TableMessagesFactory();
+
+    /** {@link SnapshotStorage} instance for the partition. */
+    private final PartitionSnapshotStorage snapshotStorage;
+
+    /** Snapshot URI. */
+    private final SnapshotUri snapshotUri;
+
+    /** Rebalance thread-pool, used to write data into a storage. */
+    //TODO Use external pool.
+    private final ExecutorService threadPool = Executors.newSingleThreadExecutor();
+
+    /**
+     * Snapshot meta read from the leader.
+     *
+     * @see SnapshotMetaRequest
+     */
+    private SnapshotMeta snapshotMeta;
+
+    /**
+     * Constructor.
+     *
+     * @param snapshotStorage Snapshot storage.
+     * @param snapshotUri Snapshot URI.
+     */
+    public IncomingSnapshotCopier(PartitionSnapshotStorage snapshotStorage, SnapshotUri snapshotUri) {
+        this.snapshotStorage = snapshotStorage;
+        this.snapshotUri = snapshotUri;
+    }
+
+    @Override
+    public void cancel() {
+        //TODO Implement.
+    }
+
+    @Override
+    public void join() throws InterruptedException {
+        //TODO Implement proper join.
+        threadPool.awaitTermination(1, TimeUnit.SECONDS);
```

Comment:
How does this work? Do we call `join` after `close`?

Reply:
No, the other way around

Reply:
Hm, does `awaitTermination` wait for any tasks to complete even if the shoutdown request was not issued?

Reply:
It's a placeholder, I should explicitly state that in the code, or simply remove it